include/irr/asset/IAsset.h

```diff
@@ -175,17 +175,23 @@ class IAsset : virtual public core::IReferenceCounted
 
 		bool restoreFromDummy(IAsset* _other, uint32_t _levelsBelow)
 		{
-			if (!canBeRestoredFrom(_other))
+			assert(getAssetType() == _other->getAssetType());
+
+			if (!canBeRestoredFrom_recurseDAG(_other))
 				return false;
 
 			restoreFromDummy_impl(_other, _levelsBelow);
 			return true;
 		}
 
-		virtual bool canBeRestoredFrom(const IAsset* _other) const
+		bool canBeRestoredFrom(const IAsset* _other) const
 		{
+			assert(getAssetType() == _other->getAssetType());
+
 			if (getMutability() != EM_MUTABLE)
 				return false;
+			if (_other->getMutability() != EM_MUTABLE)
+				return false;
 			if (!isADummyObjectForCache())
 				return false;
 			if (_other->isADummyObjectForCache())
```

Comment:
needs a call to `canBeRestoredFrom_recurseDAG` at the end

Reply:
but the `restoreFromDummy` should not call `canBeRestoredFrom` but `canBeRestoredFrom_this_node` or something like that

Reply:
> needs a call to canBeRestoredFrom_recurseDAG at the end

what, no, why

Reply:
what i've noted from our conversation:
* `restoreFromDummy` first calls `canBeRestoredFrom_recurseDAG` and early exits if false, otherwise calls `restoreFromDummy_impl`
* `restoreFromDummy_impl` calls `canBeRestoredFrom` and restores if true, but regardless of result calls `restoreFromDummy` on dependencies

Reply:
What you're missing that is the public `canBeRestoredFrom` checks for whether mutability and dummy state of self and this is ok (what I suggest calling `canBeRestoredFrom_this_node`, which would be equivalent to your current 4 if-statements) AND THEN the "matching" of the DAG

because of this, the `restoreFromDummy_impl` should not call this function to prevent the resotration of current asset in the DAG, but a protected `canBeRestoredFrom_this_node`.

`restoreFromDummy` shouldn't call the public function either but only `canBeRestoredFrom_recurseDAG` to early out.

Reply:
soo
> needs a call to canBeRestoredFrom_recurseDAG at the end
is not the case
You description above perfectly matches what actually is now

Reply:
no its not, there's one thing missing

The fucntion here, `canBeRestoredFrom`, should do the 4 if statements plus a call to `canBeRestoredFrom_recurseDAG`

Reply:
ugh I said it wrong again.